torch/_dynamo/compiled_autograd.py

```diff
@@ -183,3 +196,108 @@ def disable():
     prior = torch._C._dynamo.compiled_autograd.set_autograd_compiler(None)
     yield
     torch._C._dynamo.compiled_autograd.set_autograd_compiler(prior)
+
+
+# Invoke backward hook is a higher order op meant for both invoking a bound hook,
+# and for registering it as a call_function in the backward graph.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the hook as needed. This, in turn, means we can support hooks in backward with complex python
+# state mutation. If we were to not do this, the hooks would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
+def _invoke_in_backward(*args, fn, reenter):
+    return _invoke_in_backward_op(*args, fn=fn, reenter=reenter)
+
+
+_invoke_in_backward_op = HigherOrderOperator("_invoke_in_backward")
+
+
+def dynamo_interceding_fn_wrapper(*args, fn):
+    # This wrapper intercepts calls to fn, and calls the real fn via _invoke_in_backward
+    # However, as reenter is set to false, the call_function created during trace
+    # will point to the actual fn, and not to this function.
+    return _invoke_in_backward_op(*args, fn=fn, reenter=False)
+
+
+@_invoke_in_backward_op.py_impl(ProxyTorchDispatchMode)
+def inner_trace(*args, fn, reenter):
+    import torch._functorch.aot_autograd
+
+    assert len(args) == 1
+    grad = args[0]
+    assert isinstance(grad, torch.Tensor)
+
+    mode = _get_current_dispatch_mode()
+    if isinstance(fn, functools.partial):
+        fn.__name__ = fn.func.__name__  # type: ignore[attr-defined]
+    original_fn = fn
+    if reenter:
+        # If the reenter flag is set, we wrap the original fn in dynamo_interceding_fn_wrapper
+        # and write that to the graph. This produces an aot_autograd graph during backwards that
+        # points to dynamo_interceding_fn_wrapper. Then, during compiled autograd, we use
+        # dynamo_interceding_fn_wrapper to _invoke_in_backward the original fn under dynamo. The actual
+        # dynamo part of dynamo_interceding_fn_wrapper happens during compiled autograd.
+        fn = functools.partial(dynamo_interceding_fn_wrapper, fn=fn)
+        fn.__name__ = fn.func.__name__
+
+    is_func = torch._is_functional_tensor(grad)
+    if is_func:
+        grad = torch._functorch.aot_autograd.from_fun(grad)
+
+    proxy_args = pytree.tree_map(mode.tracer.unwrap_proxy, (grad,))
+    out_proxy = mode.tracer.create_proxy(
+        "call_function", fn, proxy_args, {}, name="invocation"
+    )
+    grad = track_tensor_tree(grad, out_proxy, constant=None, tracer=mode.tracer)
+
+    # We have a little shortcut here, wherein we DO NOT yet run a meta func, and so
+    # we take on an assumption that input and output meta matches. As such, we must introduce
+    # a runtime assert
```

Comment:
If you actually want to land this, you're going to have to explain what the eventual plan around 'meta funcs' here is.